core/src/main/scala/org/apache/spark/TaskEndReason.scala

```diff
@@ -212,8 +212,8 @@ case object TaskResultLost extends TaskFailedReason {
  * Task was killed intentionally and needs to be rescheduled.
  */
 @DeveloperApi
-case object TaskKilled extends TaskFailedReason {
-  override def toErrorString: String = "TaskKilled (killed intentionally)"
+case class TaskKilled(reason: String) extends TaskFailedReason {
+  override def toErrorString: String = s"TaskKilled ($reason)"
```

Comment:
This is unfortunately not backwards compatible. I've looked into this, but the issue seems to be that case objects are not equal to any case class in scala. If `TaskKilled` was a case class to start with, compatibility might have been possible.